### ISSUE TITLE
"Run-Ahead" doesn't really work with Beetle and the games Driver/Driver 2

### ISSUE THREAD

**Issue by iRng-ICEMAN** [libretro/beetle-psx-libretro#625]
# First and foremost consider this:
- This is a DEVELOPER oriented channel dedicated to report bugs and suggest features about the libretro core, please do not ask for help. If you need help, go to https://forums.libretro.com/;
- The upstream bug tracker is a forum thread, located at https://forum.fobby.net/index.php?t=msg&th=1114;
    - Note: As of October 2018, Microsoft email providers (think email addresses ending with hotmail/live/outlook) are rejecting email from the forum. If you want to register an account in the upstream project's forum, consider using another mail provider.
- For game-related bugs:
    1) First, check if it's already reported on the upstream bug tracker;
    2) If it's not reported there, try to reproduce the bug with mednafen;
    3) If it's reproducible with mednafen, report it on the upstream bugtracker;
    4) If it's not reproducible with mednafen, ONLY THEN should you report it here.
    5) DO NOT report Beetle-specific bugs on the upstream tracker.

## Description
The "RunAhead" Latency feature is quite nice and i appreciate your work, but i would like to submit this issue regarding the Beetle core and the games Driver and Driver 2.
When turning on RunAhead (doesnt matter how many frames) the FPS is trash, so much that its easily considered unplayable, like 1-5 fps or something.

I tried different core options, different settings but doesnt seem to work, atleast not with the settings i had.
[Description of the bug]

### Expected behavior
No Latency and good performance.
[What you expected to happen]
[Photos/videos of original hardware behavior and/or mednafen behavior]

### Actual behavior
No Latency (checked with "Pause, press K once to see if the vehicle moves) but very bad performance, unplayable.
[What is actually happening]
[Photos/videos of the core's behavior]

### Steps to reproduce the bug

1. Turn on RunAhead
2. Set frames (i tried between 1 and 8, 5 frames for Driver 2 and 7/8 frames for Driver 1)
3. Continue the game and see horrible performance

### Bisect results

[Try to bisect and tell us when this started happening]

### Version/Commit
Retroarch: 1.8.4
[If you're using RetroArch, this information is on the bottom left after loading the core. You can do this via Main Menu -> Load Core]

- Beetle PSX: [version/commit] or Beetle PSX HW: [version/commit]
Beetle PSX HW/0.9.44.1 (tried with Beetle PSX aswell and same issue)
Please specify whether you are using Beetle PSX or Beetle PSX HW.

### Core options
Driver 2 USA
Renderer: Hardware (Auto)
Software Framevuffer: On
Internal GPU Resolution: 4x
Internal Color Depth: 32bpp
Dithering Pattern: Internal Resolution
Texture Filtering: Bilinear
Wireframe Mode: Off
Display Full VRAM: Off
PGXP Operation Mode: Memory + CPU (Buggy)
PGXP Vertex Cache: Off
PGXP Perspective Correct Texturing: On
Display Internal FPS: Off
Line-to-Quad Hack: Default
CPU Dynarec: Max Performance
Dynarec Code Invalidation: Full
Dynarec DMA/GPU Event Cycles: 1024
CPU Frequency Scaling (Overclock): 200%
GTE Overlock: On
GPU Rastersizer Overclock: 32x
Skip Bios: On
Core-Reported FPS Timing: Progressive Rate (Default)
Core Aspect Ratio: Corrected
Widescreen Mode Hack: On
Crop Horizontal Overscan: On
Horizontal Image Offset: 0
Initial Scanline - NTSC: 0
Last Scanline - NTSC: 239
Initial Scanline - Pal: 0
Last Scanline - Pal: 287
CD Access Method: Asynchronous
CD Loading Speed: 4x
Memory Card 0 Method: Libretro
Enable Memory Card 1: On
Shared Memcards: Off
Analog Self-Calibration: Off
Enable Dualshock Analog Mode: Off
Port 1: Multitap Enable: Off
Port 2: Multitap Enable: Off
Gun Input Mode: Light Gun
Gun Cursor: Cross
Mouse Sentivitiy: 100%
NegCon Twist Responce: Linear
NegCon Twist Deadzone: 0%

Latency:
Hard GPU Sync: On
Hard GPU Sync Frames: 3
Frame Delay: 0
Audi Latency (ms): 64
Poll Type Behavior: Late
Run-Ahead to Reduce Latency: On
Frames: 1-8
[List all the core options you have changed]

Please also specify, if reporting a graphical glitch for Beetle PSX HW, your "Renderer" core option setting and your frontend selected video driver.

### Environment information

- Frontend: [Frontend name and version]
- OS: Win7
- Compiler: [In case you are running local builds]
- GPU: 980ti

**Comment by twinaphex** [comment]
Runahead doesnt work reliably with the Hardware renderer. You need to use the Software renderer instead.

**Comment by iRng-ICEMAN** [comment]
> 
> 
> Runahead doesnt work reliably with the Hardware renderer. You need to use the Software renderer instead.

Like i said, i used
- beetle hw in Hardware mode
- beetle hw in Software mode
- beetle (non hw version)

So, what exactly should i try?... these 3 didnt help.

**Comment by ZachCook** [comment]
Does it work better with dynarec off? Runahead clears some of the dynarec's state, does the runahead second instance option work any better either?

**Comment by iRng-ICEMAN** [comment]
> Does it work better with dynarec off? Runahead clears some of the dynarec's state, does the runahead second instance option work any better either?

Nope, doesnt help either. I hope the Devs will have a look into this, quite disappointing to not be able to run this with modern machines.

**Comment by ZachCook** [comment]
You are running with 4x software (cpu) rendering (due to the software framebuffer option), PGXP and Runahead, all of which drastically increase requirements on the cpu, I can also see a decrease in performance from 270 fps to 20 fps in driver 2 when running with those options.

Try disabling PGXP and software framebuffer, and it should run much better when using runahead.

**Comment by ggdrt** [comment]
@ZachCook Does the core not fall back properly by default to the software renderer at 1x IR? If so then I'll need to take a look at that section of code again.

**Comment by ZachCook** [comment]
@ggdrt I have no idea if it does or not, just guessing at why it is so slow with the listed settings, as I know software renderer is really slow at 4xIR + PGXP + runahead

**Comment by ggdrt** [comment]
Ah sorry, I misunderstood your comment.

I think it primarily comes down to run-ahead being extremely taxing when running Beetle PSX. Beetle PSX is single-threaded and 4x software rendering is pretty much the limit for midrange Intel i5 processors in the last couple of generations (9th gen and earlier). Runahead means that for every frame the user sees, we have to serially emulate X additional runahead frames within that one-frame period, which most users won't have enough cpu time for at 4x IR.

**Comment by ggdrt** [comment]
Ah sorry, I misunderstood your comment.

I think it primarily comes down to run-ahead being extremely taxing when running Beetle PSX. Beetle PSX is single-threaded and 4x software rendering is pretty much the limit for midrange Intel i5 processors in the last couple of generations (9th gen and earlier). Runahead means that for every frame the user sees, we have to emulate X additional runahead frames within that one-frame period, which most users won't have enough cpu time for at 4x IR.

**Comment by ggdrt** [comment]
Ah sorry, I misunderstood your comment.

I think it primarily comes down to run-ahead being extremely taxing when running Beetle PSX. Beetle PSX is single-threaded and 4x software rendering is pretty much the limit for midrange Intel i5 processors in the last couple of generations (9th gen and earlier). Runahead means that for every frame the user sees, we have to emulate X additional runahead frames within that one-frame period, which most users won't have enough cpu time for at 4x IR.

**Comment by iRng-ICEMAN** [comment]
@ZachCook @ggdrt 

I tried 3 different setups and none of them worked well.

I have the feeling one of them could work well but the issue is i dont get a output. I mean, when changing the settings and combining like 1x, PGXP mode disabled, softwarebuffer disabled, runahead (5) + sound fix -> i cant get a image. I can hear the game in the backround running well but no picture. It could be bugging with Retroarch maybe i dont know.

**Comment by ggdrt** [comment]
If you're able to toggle the software framebuffer core option, that means you're running Beetle in hardware rendering mode. As twinaphex mentioned previously, you should use the software rendering mode instead.

Even without runahead, Beetle already has high performance requirements. If you're trying to do 5 frames of runahead, you're going to run into problems unless you have a processor with incredible single thread performance.
